compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp

```diff
@@ -0,0 +1,622 @@
+//
+//===----------------------------------------------------------------------===//
+
+#include "iree/compiler/Dialect/Flow/Conversion/TensorToFlow/Utils.h"
+#include "iree/compiler/Dialect/Flow/Transforms/RegionOpUtils.h"
+#include "iree/compiler/GlobalOptimization/PassDetail.h"
+#include "iree/compiler/GlobalOptimization/Passes.h"
+#include "llvm/Support/Debug.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/Dialect/Linalg/Transforms/Transforms.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/Dialect/Utils/IndexingUtils.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+#define DEBUG_TYPE "iree-opt-propagate-linalg-transpose"
+
+namespace mlir::iree_compiler::GlobalOptimization {
+
+//===----------------------------------------------------------------------===//
+// Transpose permutation helpers
+//===----------------------------------------------------------------------===//
+
+static bool isIdentityPermutation(ArrayRef<int64_t> perm) {
+  for (auto [index, dim] : llvm::enumerate(perm)) {
+    if (index != dim) {
+      return false;
+    }
+  }
+  return true;
+}
+
+static bool arePermutationsEqual(ArrayRef<int64_t> first,
+                                 ArrayRef<int64_t> second) {
+  if (first.size() != second.size()) {
+    return false;
+  }
+  return llvm::all_of(llvm::zip_equal(first, second),
+                      [](std::tuple<int64_t, int64_t> p) {
+                        return std::get<0>(p) == std::get<1>(p);
+                      });
+}
+
+// Constructs a transpose of the given tensor and permutation.
+static Value createTranspose(OpBuilder &builder, Value source,
+                             ArrayRef<int64_t> perm) {
+  SmallVector<OpFoldResult> mixedSizes =
+      tensor::getMixedSizes(builder, source.getLoc(), source);
+  applyPermutationToVector(mixedSizes, perm);
+  Type elemType = cast<RankedTensorType>(source.getType()).getElementType();
+  Value empty =
+      builder.create<tensor::EmptyOp>(source.getLoc(), mixedSizes, elemType)
+          .getResult();
+  return builder
+      .create<linalg::TransposeOp>(source.getLoc(), source, empty, perm)
+      ->getResult(0);
+}
+
+static RankedTensorType getPermutedTensorType(RankedTensorType type,
+                                              SmallVector<int64_t> perm) {
+  SmallVector<int64_t> permutedShape = applyPermutation(type.getShape(), perm);
+  return RankedTensorType::get(permutedShape, type.getElementType());
+}
+
+//===----------------------------------------------------------------------===//
+// Transpose specialization
+//===----------------------------------------------------------------------===//
+
+// Indicates whether the given linalg op represents a transpose. In particular,
+// it requires a single input where the indexing maps are full permutations and
+// non-equal.
+static bool isaTransposeOpInterface(linalg::LinalgOp linalgOp) {
```

Comment:
yes please :)

I can move this upstream now if that's preferred. I just didn't want to have to wait for upstream changes to land and roll in, but I'm not in a rush anymore. If we landed this as-is, I'd move it upstream as a follow up anyway.

Reply:
FYI - We were discussing having a `isTranspose(linalg::LinalgOp op)` utility during [this patch's review](https://github.com/llvm/llvm-project/pull/74206#pullrequestreview-1765733555) (see point 1). 

Once upstreamed, this will also benefit those changes :D

Reply:
Currently this is run after `FoldUnitExtentDims` to avoid extra handling for such cases, but when an upstream interface is added we can reconsider such variations. Also there is an additional limitation in that we are only propagating "true" transposes here (i.e. full rank) as this is before fusion of generics.